Updated metadata documentation part 2 - principles

bjwheltor opened the pull request:
Description

An initial draft of the Principles metadata document (a re-working of information currently hosted on an internal Met Office site). This creates a Principles.rst file.

gavinevans commented on doc/source/metadata/Principles.rst:
```diff
@@ -0,0 +1,208 @@
+Clear purpose
+-------------
+
+Metadata should not be present, if it does not have a purpose
+(e.g. it should be included just because it might be useful in future), as,
+if no one is using it, it is unlikely to be properly maintained.
```
I thought that this might be clearer.

```suggestion
Metadata should only be present if it has a purpose. It should not 
be included just because it might be useful in future, as if no one is 
using it, it is unlikely to be properly maintained.
```

bjwheltor replied:
Agreed - making change.

gavinevans commented on doc/source/metadata/Principles.rst:
```diff
@@ -0,0 +1,208 @@
+Nothing misleading or unnecessarily restrictive
+-----------------------------------------------
+
+Metadata that misleading, or simply wrong, should not be present.
```
Typo:

```suggestion
Metadata that is misleading, or simply wrong, should not be present.
```

gavinevans commented on doc/source/metadata/Principles.rst:
```diff
@@ -0,0 +1,208 @@
+(the 'where' metadata described in the earlier note).
+
+One approach to help support this that has been partially adopted within
+IMPROVER is the use of such as name-spacing for new (non-CF) attributes.
```
Typo:

```suggestion
IMPROVER is the use name-spacing for new (non-CF) attributes.
```

bjwheltor replied:
Thanks - corrected.

gavinevans commented on doc/source/metadata/Principles.rst:
```diff
@@ -0,0 +1,208 @@
+
+ * It clearly separates the attribute from CF standard attributes,
+   identifying it as belonging to a separate metadata convention.
+ * It provides the potential this to be 'resolved' to a definition
```
Perhaps this should be like this?

```suggestion
 * It provides the potential for this attribute to be 'resolved' to a definition
```

bjwheltor replied:
Corrected.

gavinevans commented on doc/source/metadata/Principles.rst:
```diff
@@ -0,0 +1,208 @@
+ 
+At present, IMPROVER does not provide explicit references to external metadata.
+Earlier, in the development process,the use of NetCDF Linked Data was explored
+(see note below), but the method of implementation a the time created issues
```
Typo:

```suggestion
(see note below), but the method of implementation at the time created issues
```

bjwheltor replied:
Corrected.

gavinevans commented on doc/source/metadata/Principles.rst:
```diff
@@ -0,0 +1,208 @@
+
+IMPROVER code is usually implemented in a series of processing chains,
+so it makes sense to consider the metadata in this context. 
+From a post-processing perspective, a helpful to divide the metadata 
```
Typo:

```suggestion
From a post-processing perspective, it is helpful to divide the metadata 
```

bjwheltor replied:
corrected

gavinevans commented on doc/source/metadata/Principles.rst:
```diff
@@ -0,0 +1,208 @@
+    (delete, set)
+
+The general approach proposed is to be conservative with metadata;
+remove attributed on the source data that do not serve a purpose for the
```
Typo:

```suggestion
remove attributes on the source data that do not serve a purpose for the
```

bjwheltor replied:
corrected

gavinevans commented on doc/source/metadata/Principles.rst:
```diff
@@ -0,0 +1,208 @@
+Organisation-specific metadata may be added in at the end of the processing chain.
+
+Low-level metadata will usually only be transitory,
+required for certain processing steps, but out exposed in the final output.
```
Typo:

```suggestion
required for certain processing steps, but not exposed in the final output.
```

bjwheltor replied:
Thanks, corrected.

gavinevans reviewed (changes requested):
Thanks @bjwheltor 👍 

I've added a few minor comments to `Principles.rst`, mostly to correct typos. Rebasing this PR on master should help to get rid of the majority of the commits associated with https://github.com/metoppv/improver/pull/1846.

bjwheltor commented:
> Thanks @bjwheltor 👍
> 
> I've added a few minor comments to `Principles.rst`, mostly to correct typos. Rebasing this PR on master should help to get rid of the majority of the commits associated with #1846.

@gavinevans - I've addressed your comments - thanks.
I did try rebasing on master, but I got:

```
Patch failed at 0001 Initial structure for metadata documentation, very limited content
hint: Use 'git am --show-current-patch' to see the failed patch
```

I had a look `git am`, but I'm not really sure what I'm doing and backed out in the end rather than leave the branch in a messy state.

bjwheltor commented:
@gavinevans - thanks for the help, 're-basing' it, have now tidied it up - just the 4 commits!

brhooper commented on doc/source/metadata/Principles.rst:
```diff
@@ -0,0 +1,208 @@
+
+Namespacing is implemented using a double underscore to maintain CF conformance
+whilst clearly identifying the namespace part of the attribute name.
+For example, the "ssp" is used as statistical post-processing namespace,
```
```suggestion
For example, "ssp" is used to denote the statistical post-processing namespace,
```

bjwheltor replied:
Done

brhooper commented on doc/source/metadata/Principles.rst:
```diff
@@ -0,0 +1,208 @@
+
+The other namespace used is "mosg", which is used to indicate
+a Met Office standard grid attribute,
+which despite the name can, and should, generally be used.
```
This is currently quite a long sentence, the following edit might aid readability.

```suggestion
Despite the name this can, and generally should, be used.
```

bjwheltor replied:
Done

brhooper reviewed (changes requested):
A few very small review comments to fix typos and re-word a couple of sentences.

bjwheltor commented on doc/source/metadata/Principles.rst:
```diff
@@ -0,0 +1,208 @@
+
+As far as possible the use of metadata within IMPROVER should conform to the
+`CF Metadata Conventions`_ (CF),
+but recognising the need to extend this in some areas.
```
done

bjwheltor commented on doc/source/metadata/Principles.rst:
```diff
@@ -0,0 +1,208 @@
+However, whilst CF does a good job with 'where' and 'what' metadata
+(see note below), it does not really cover 'who' metadata.
+Therefore, it is helpful to keep CF (and direct extensions to this)
+focussed on these aspects and not try to stretch it to cover 'who'.
```
Happy to change - I've never sure and it reduces the word count  :smile:

bjwheltor commented on doc/source/metadata/Principles.rst:
```diff
@@ -0,0 +1,208 @@
+An example is CF cell methods, which are intended to cover 'what'
+(i.e. to ensure that data can be correctly interpreted) 
+but can unwittingly be extended into 'who'
+(e.g. to try to describe the post-processing that have been applied), 
```
Done

bjwheltor commented:
@brhooper - thanks for your comments. I've made all the changes and updated the PR, so it should probably be okay to merge now.

brhooper reviewed (approved):
Thanks for those changes @bjwheltor, all looks good
